iox_arrow_flight/src/encode.rs

```diff
@@ -0,0 +1,375 @@
+    }
+
+    /// Include `app_metadata` in the first Schema message
+    pub fn with_metadata(mut self, app_metadata: Vec<u8>) -> Self {
+        self.app_metadata = app_metadata;
+        self
+    }
+
+    /// With updated options
+    pub fn with_options(mut self, options: IpcWriteOptions) -> Self {
+        self.options = options;
+        self
+    }
+
+    /// Return a stream that converts a  [`Stream`](futures::Stream) of
+    /// [`RecordBatch`] to a stream of [`FlightData`], consuming self.
+    pub fn build(
+        self,
+        schema: SchemaRef,
+        input: BoxStream<'static, Result<RecordBatch, tonic::Status>>,
+    ) -> BoxStream<'static, Result<FlightData, tonic::Status>> {
+        let Self {
+            max_batch_size,
+            app_metadata,
+            options,
+        } = self;
+
+        // The first message is the schema message, and all batches have
+        // to have that schema too
+        let schema = Arc::new(prepare_schema_for_flight(&schema));
+        let mut schema_flight_data: FlightData = SchemaAsIpc::new(&schema, &options).into();
+        schema_flight_data.app_metadata = app_metadata;
+
+        let schema_stream = futures::stream::once(async move { Ok(schema_flight_data) });
+
+        let options = Arc::new(options);
+
+        let data_stream = input
+            .map(move |r| {
+                let options = Arc::clone(&options);
+                let schema = Arc::clone(&schema);
+                async move {
+                    match r {
+                        Err(e) => futures::stream::iter(vec![Err(e)]).boxed(),
+                        Ok(batch) => encode_batch(schema, max_batch_size, batch, &options),
+                    }
+                }
+            })
+            // buffer this conversion so there can be  record batches can be
+            // produced while others are sent over the network
+            .buffered(1)
+            .flatten();
+
+        // Send the schema first then the data
+        schema_stream.chain(data_stream).boxed()
+    }
+}
+
+/// Encodes  batch as one or more FlightData structures
+fn encode_batch(
+    schema: SchemaRef,
+    max_batch_size: usize,
+    batch: RecordBatch,
+    options: &IpcWriteOptions,
+) -> BoxStream<'static, Result<FlightData, tonic::Status>> {
+    let batch = match prepare_batch_for_flight(&batch, schema) {
+        Ok(batch) => batch,
+        Err(e) => {
+            let e = tonic::Status::internal(format!(
+                "Preparing to send batch over Arrow Flight: {}",
+                e
+            ));
+            return futures::stream::iter(vec![Err(e)]).boxed();
+        }
+    };
+
+    let mut output = vec![];
+    for batch in split_batch_for_grpc_response(batch, max_batch_size) {
+        let (flight_dictionaries, flight_batch) = flight_data_from_arrow_batch(&batch, options);
+        output.extend(flight_dictionaries);
+        output.push(flight_batch)
+    }
+    let output = output.into_iter().map(Ok);
```

Comment:
Feel free to punt this until profiling to know if this makes a difference, but I think some allocations can be saved with:

```suggestion
    let output: Vec<_> = split_batch_for_grpc_response(batch, max_batch_size)
        .into_iter()
        .map(|batch| {
            let (flight_dictionaries, flight_batch) = flight_data_from_arrow_batch(&batch, options);
            flight_dictionaries
                .into_iter()
                .chain(std::iter::once(flight_batch))
        })
        .flatten()
        .map(Ok)
        .collect();
```